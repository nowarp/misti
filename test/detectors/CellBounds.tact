struct S {
    a: Int;
    b: Int;
}

fun refs_underflow(c: Cell) {
    let s1 = beginCell()
               .storeRef(c)
               .endCell()
               .asSlice();
    let ref1 = s1.loadRef(); // OK
    let ref2 = s1.loadRef(); // Bad: Cell Underflow

    beginCell() // Bad: Cell Underflow
      .endCell()
      .asSlice()
      .loadRef();

    // Test structures/messages
    let st1 = S {a: 1 , b: 2};
    let c1 = st1.toCell().asSlice().loadRef(); // Bad: Cell Underflow
    st1.toCell().asSlice().loadRef(); // Bad: Cell Underflow
}

fun refs_overflow(c: Cell) {
    // Cell Overflow when storing a known local builder
    let b1 = beginCell()
               .storeRef(c)
               .storeRef(c)
               .storeRef(c)
               .storeRef(c); // OK: Builder storing 4 references
    let c1 = beginCell() // Bad: We got 5 references
               .storeBuilder(b1)
               .storeRef(c)
               .endCell();

    let c2 = // Bad: 4+ refs
    beginCell()
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .endCell();

   let b2 = beginCell();
   b2 = b2.storeRef(c);
   b2 = b2.storeRef(c);
   b2 = b2.storeRef(c);
   b2 = b2.storeRef(c);
   b2 = b2.storeRef(c); // Bad: 4+ refs
   let c3 = b2.endCell(); // Bad: 4+ refs

   let b3 = b2;
   let c4 = b3.endCell(); // Bad: 4+ refs
}

fun data_overflow() {
    let b1 = beginCell();
    b1 = b1.storeInt(42, 257);
    b1 = b1.storeInt(42, 259-2);
    b1 = b1.storeInt(42, 256+1);
    b1 = b1.storeInt(42, 257); // Bad: 1023+ bits
    let c1 = b1.endCell(); // Bad: 1023+ bits

    let b2 = beginCell();
    b2 = b2.storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 210); // Total size: 981
    b2 = b2.storeCoins(40000000000); // Bad: 44 + 981 = 1025 => overflow
    c1 = b2.endCell(); // Bad

    let b_1000_bits = beginCell()
           .storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 257)
           .storeInt(1, 229);
    let b3 = beginCell();
    b3 = b3.storeBuilder(b_1000_bits) // Bad: 1025 bits
           .storeInt(1, 25);
    c1 = b3.endCell(); // Bad: 1025 bits
}


fun refs_num_ok(c: Cell, copt: Cell?) {
    let c1 = // OK: Cell could be None (no taint analysis yet)
    beginCell()
        .storeRef(c)
        .storeRef(c)
        .storeRef(c)
        .storeMaybeRef(copt)
        .storeRef(c)
        .endCell();

    // OK: no assignments to the result of storeRef calls
    let b1 = beginCell();
    b1.storeRef(c);
    b1.storeRef(c);
    b1.storeRef(c);
    b1.storeRef(c);
    b1.storeRef(c);
    c1 = b1.endCell();
}
