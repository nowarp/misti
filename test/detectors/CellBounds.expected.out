[CRITICAL] CellBounds: Reference count might go below 0
test/detectors/CellBounds.tact:7:9:
  6 | fun refs_underflow(c: Cell) {
> 7 |     let s1 = beginCell()
              ^
  8 |                .storeRef(c)
The possible number of references stored (1) is less than loaded (2)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Reference count might go below 0
test/detectors/CellBounds.tact:12:9:
  11 |     let ref1 = s1.loadRef(); // OK
> 12 |     let ref2 = s1.loadRef(); // Bad: Cell Underflow
               ^
  13 | 
The possible number of references stored (1) is less than loaded (2)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Reference count might go below 0
test/detectors/CellBounds.tact:14:5:
  13 | 
> 14 |     beginCell() // Bad: Cell Underflow
           ^
  15 |       .endCell()
The possible number of references stored (0) is less than loaded (1)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Reference count might go below 0
test/detectors/CellBounds.tact:21:9:
  20 |     let st1 = S {a: 1 , b: 2};
> 21 |     let c1 = st1.toCell().asSlice().loadRef(); // Bad: Cell Underflow
               ^
  22 |     st1.toCell().asSlice().loadRef(); // Bad: Cell Underflow
The possible number of references stored (0) is less than loaded (2)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Reference count might go below 0
test/detectors/CellBounds.tact:22:5:
  21 |     let c1 = st1.toCell().asSlice().loadRef(); // Bad: Cell Underflow
> 22 |     st1.toCell().asSlice().loadRef(); // Bad: Cell Underflow
           ^
  23 | }
The possible number of references stored (0) is less than loaded (1)
Help: Remove extra .loadRef operations or store more refs first
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Too many references stored in cell
test/detectors/CellBounds.tact:32:9:
  31 |                .storeRef(c); // OK: Builder storing 4 references
> 32 |     let c1 = beginCell() // Bad: We got 5 references
               ^
  33 |                .storeBuilder(b1)
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Too many references stored in cell
test/detectors/CellBounds.tact:37:9:
  36 | 
> 37 |     let c2 = // Bad: 4+ refs
               ^
  38 |     beginCell()
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Too many references stored in cell
test/detectors/CellBounds.tact:51:4:
  50 |    b2 = b2.storeRef(c);
> 51 |    b2 = b2.storeRef(c); // Bad: 4+ refs
          ^
  52 |    let c3 = b2.endCell(); // Bad: 4+ refs
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Too many references stored in cell
test/detectors/CellBounds.tact:54:8:
  53 | 
> 54 |    let b3 = b2;
              ^
  55 |    let c4 = b3.endCell(); // Bad: 4+ refs
The minimum number of references stored (5) exceeds the maximum allowed (4)
Help: Split your data across multiple cells - a single cell cannot store more than 4 references
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Data size exceeds cell capacity
test/detectors/CellBounds.tact:63:5:
  62 |     b1 = b1.storeInt(42, 256+1);
> 63 |     b1 = b1.storeInt(42, 257); // Bad: 1023+ bits
           ^
  64 |     let c1 = b1.endCell(); // Bad: 1023+ bits
The minimum data size stored (1028 bits) exceeds the maximum allowed (1023 bits)
Help: Split your data across multiple cells - a single cell cannot store more than 1023 bits
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Data size exceeds cell capacity
test/detectors/CellBounds.tact:71:5:
  70 |            .storeInt(1, 210); // Total size: 981
> 71 |     b2 = b2.storeCoins(40000000000); // Bad: 44 + 981 = 1025 => overflow
           ^
  72 |     c1 = b2.endCell(); // Bad
The minimum data size stored (1025 bits) exceeds the maximum allowed (1023 bits)
Help: Split your data across multiple cells - a single cell cannot store more than 1023 bits
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds

[CRITICAL] CellBounds: Data size exceeds cell capacity
test/detectors/CellBounds.tact:80:5:
  79 |     let b3 = beginCell();
> 80 |     b3 = b3.storeBuilder(b_1000_bits) // Bad: 1025 bits
           ^
  81 |            .storeInt(1, 25);
The minimum data size stored (1025 bits) exceeds the maximum allowed (1023 bits)
Help: Split your data across multiple cells - a single cell cannot store more than 1023 bits
See: https://nowarp.io/tools/misti/docs/detectors/CellBounds